secboot/secboot_tpm.go

```diff
@@ -244,3 +264,166 @@ func unlockEncryptedPartition(tpm *sb.TPMConnection, name, device, keyfile, pinf
 
 	return nil
 }
+
+func SealKey(key partition.EncryptionKey, params *SealKeyParams) error {
+	numModels := len(params.ModelParams)
+	if numModels < 1 {
+		return fmt.Errorf("at least one set of model-specific parameters is required")
+	}
+
+	tpm, err := sbConnectToDefaultTPM()
+	if err != nil {
+		return fmt.Errorf("cannot connect to TPM: %v", err)
+	}
+
+	modelPCRProfiles := make([]*sb.PCRProtectionProfile, 0, numModels)
+
+	for _, modelParams := range params.ModelParams {
+		modelProfile := sb.NewPCRProtectionProfile()
+
+		// Verify if all EFI image files exist
+		for _, chain := range modelParams.EFILoadChains {
+			if err := checkFilesPresence(chain); err != nil {
+				return err
+			}
+		}
+
+		// Add EFI secure boot policy profile
+		policyParams := sb.EFISecureBootPolicyProfileParams{
+			PCRAlgorithm:  tpm2.HashAlgorithmSHA256,
+			LoadSequences: make([]*sb.EFIImageLoadEvent, 0, len(modelParams.EFILoadChains)),
+			// TODO:UC20: set SignatureDbUpdateKeystore to support key rotation
+		}
+		for _, chain := range modelParams.EFILoadChains {
+			policyParams.LoadSequences = append(policyParams.LoadSequences, buildLoadSequence(chain))
+		}
+		if err := sbAddEFISecureBootPolicyProfile(modelProfile, &policyParams); err != nil {
+			return fmt.Errorf("cannot add EFI secure boot policy profile: %v", err)
+		}
+
+		// Add systemd EFI stub profile
+		if len(modelParams.KernelCmdlines) != 0 {
+			systemdStubParams := sb.SystemdEFIStubProfileParams{
+				PCRAlgorithm:   tpm2.HashAlgorithmSHA256,
+				PCRIndex:       tpmPCR,
+				KernelCmdlines: modelParams.KernelCmdlines,
+			}
+			if err := sbAddSystemdEFIStubProfile(modelProfile, &systemdStubParams); err != nil {
+				return fmt.Errorf("cannot add systemd EFI stub profile: %v", err)
+			}
+		}
+
+		// Add snap model profile
+		if modelParams.Model != nil {
+			snapModelParams := sb.SnapModelProfileParams{
+				PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+				PCRIndex:     tpmPCR,
+				Models:       []*asserts.Model{modelParams.Model},
+			}
+			if err := sbAddSnapModelProfile(modelProfile, &snapModelParams); err != nil {
+				return fmt.Errorf("cannot add snap model profile: %v", err)
+			}
+		}
+
+		modelPCRProfiles = append(modelPCRProfiles, modelProfile)
+	}
+
+	var pcrProfile *sb.PCRProtectionProfile
+	if numModels > 1 {
+		pcrProfile = sb.NewPCRProtectionProfile().AddProfileOR(modelPCRProfiles...)
+	} else {
+		pcrProfile = modelPCRProfiles[0]
+	}
+
+	// Provision the TPM as late as possible
+	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		return err
+	}
+
+	// If we are re-installing, there will be an a sealed keyfile that already
```

Comment:
oh also we need a unit test about this deletion

Reply:
Right, will add a unit test. I placed key removal there because it's after the TPM provisioning barrier so that if TPM is successfully reprovisioned (meaning that it was cleared), then all hopes are lost of being able to use the existing sealed key and it's safe to delete it. If we delete the sealed key earlier, but then reinstall fails because the TPM is already provisioned, there's no way back and access to (potentially) existing data is lost (unless we resort to the recovery key). But I'll be happy to discuss this further and change the placement of key removal if there's an advantage in doing it elsewhere.

Reply:
ah I was just going off of code organization standpoint, but also at this point where we are provisioning the TPM we've already deleted the previous partition, so do we consider that data still recoverable by data forensics, etc. and trying to re-build the partition and use the key to unseal? It's not clear to me that's feasible, since we will have created/written new partition table and partitions on top of the old ones, but I suppose probably the data content hasn't been overwritten yet...

Anyways if this is the case I think it's fine to leave the deletion here in that case, but it still feels slightly cleaner/more obvious to do this in the other function, but then again I could see an argument that this operation is clearer/more obvious here anyways.

Reply:
Hmm yes, you're right, the partition was already deleted at this point, so my previous rationale is invalid. However, it is still useful to handle the keyfile inside secboot because other key sealing technologies may store the key in some place outside of the filesystem (the keyfile path parameter in this case would not be used). Can we think a bit more about and address it in a follow up if needed?

Reply:
yes this is fine for a follow-up and to discuss with a larger group